{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# pycoQC CLI Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "PycoQC CLI can generate a beautiful HTML formatted report containing interactive D3.js plots. On top of it, the CLI can also dump summary information in a JSON formated file allowing easy parsing with third party tools.\n",
    "\n",
    "The report is dynamically generated depending on the information available in the summary file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CLI Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Activate virtual environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-16T13:15:25.978368Z",
     "start_time": "2020-12-16T13:15:25.675755Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(pycoQC) (pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Using conda here but can also be done with other virtenv managers \n",
    "conda activate pycoQC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting help"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-16T13:15:29.434767Z",
     "start_time": "2020-12-16T13:15:27.842415Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: pycoQC [-h] [--version]\n",
      "              [--summary_file [SUMMARY_FILE [SUMMARY_FILE ...]]]\n",
      "              [--barcode_file [BARCODE_FILE [BARCODE_FILE ...]]]\n",
      "              [--bam_file [BAM_FILE [BAM_FILE ...]]]\n",
      "              [--html_outfile HTML_OUTFILE] [--json_outfile JSON_OUTFILE]\n",
      "              [--min_pass_qual MIN_PASS_QUAL] [--min_pass_len MIN_PASS_LEN]\n",
      "              [--filter_calibration] [--filter_duplicated]\n",
      "              [--min_barcode_percent MIN_BARCODE_PERCENT]\n",
      "              [--report_title REPORT_TITLE] [--template_file TEMPLATE_FILE]\n",
      "              [--config_file CONFIG_FILE] [--skip_coverage_plot]\n",
      "              [--sample SAMPLE] [--default_config] [-v | -q]\n",
      "\n",
      "pycoQC computes metrics and generates interactive QC plots from the sequencing summary\n",
      "report generated by Oxford Nanopore technologies basecallers\n",
      "\n",
      "* Minimal usage\n",
      "    pycoQC -f sequencing_summary.txt -o pycoQC_output.html\n",
      "* Including Guppy barcoding file + html output + json output\n",
      "    pycoQC -f sequencing_summary.txt -b barcoding_sequencing.txt -o pycoQC_output.html -j pycoQC_output.json\n",
      "* Including Bam file + html output\n",
      "    pycoQC -f sequencing_summary.txt -a alignment.bam -o pycoQC_output.html\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  --version             show program's version number and exit\n",
      "  -v, --verbose         Increase verbosity\n",
      "  -q, --quiet           Reduce verbosity\n",
      "\n",
      "Input/output options:\n",
      "  --summary_file [SUMMARY_FILE [SUMMARY_FILE ...]], -f [SUMMARY_FILE [SUMMARY_FILE ...]]\n",
      "                        Path to a sequencing_summary generated by Albacore\n",
      "                        1.0.0 + (read_fast5_basecaller.py) / Guppy 2.1.3+\n",
      "                        (guppy_basecaller). One can also pass multiple space\n",
      "                        separated file paths or a UNIX style regex matching\n",
      "                        multiple files (Required)\n",
      "  --barcode_file [BARCODE_FILE [BARCODE_FILE ...]], -b [BARCODE_FILE [BARCODE_FILE ...]]\n",
      "                        Path to the barcode_file generated by Guppy 2.1.3+\n",
      "                        (guppy_barcoder) or Deepbinner 0.2.0+. This is not a\n",
      "                        required file. One can also pass multiple space\n",
      "                        separated file paths or a UNIX style regex matching\n",
      "                        multiple files (optional)\n",
      "  --bam_file [BAM_FILE [BAM_FILE ...]], -a [BAM_FILE [BAM_FILE ...]]\n",
      "                        Path to a Bam file corresponding to reads in the\n",
      "                        summary_file. Preferably aligned with Minimap2 One can\n",
      "                        also pass multiple space separated file paths or a\n",
      "                        UNIX style regex matching multiple files (optional)\n",
      "  --html_outfile HTML_OUTFILE, -o HTML_OUTFILE\n",
      "                        Path to an output html file report (required if\n",
      "                        json_outfile not given)\n",
      "  --json_outfile JSON_OUTFILE, -j JSON_OUTFILE\n",
      "                        Path to an output json file report (required if\n",
      "                        html_outfile not given)\n",
      "\n",
      "Filtering options:\n",
      "  --min_pass_qual MIN_PASS_QUAL\n",
      "                        Minimum quality to consider a read as 'pass' (default:\n",
      "                        7)\n",
      "  --min_pass_len MIN_PASS_LEN\n",
      "                        Minimum read length to consider a read as 'pass'\n",
      "                        (default: 0)\n",
      "  --filter_calibration  If given, reads flagged as calibration strand by the\n",
      "                        basecaller are removed (default: False)\n",
      "  --filter_duplicated   If given, duplicated read_ids are removed but the\n",
      "                        first occurence is kept (Guppy sometimes outputs the\n",
      "                        same read multiple times) (default: False)\n",
      "  --min_barcode_percent MIN_BARCODE_PERCENT\n",
      "                        Minimal percent of total reads to retain barcode\n",
      "                        label. If below, the barcode value is set as\n",
      "                        `unclassified` (default: 0.1)\n",
      "\n",
      "HTML report options:\n",
      "  --report_title REPORT_TITLE\n",
      "                        Title to use in the html report (default: PycoQC\n",
      "                        report)\n",
      "  --template_file TEMPLATE_FILE\n",
      "                        Jinja2 html template for the html report (default: )\n",
      "  --config_file CONFIG_FILE\n",
      "                        Path to a JSON configuration file for the html report.\n",
      "                        If not provided, looks for it in ~/.pycoQC and\n",
      "                        ~/.config/pycoQC/config. If it's still not found,\n",
      "                        falls back to default parameters. The first level keys\n",
      "                        are the names of the plots to be included. The second\n",
      "                        level keys are the parameters to pass to each plotting\n",
      "                        function (default: )\")\n",
      "  --skip_coverage_plot  Skip the coverage plot in HTML report. Useful when\n",
      "                        using a reference file containing many sequences, i.e.\n",
      "                        transcriptome (default: False)\n",
      "\n",
      "Other options:\n",
      "  --sample SAMPLE       If not None a n number of reads will be randomly\n",
      "                        selected instead of the entire dataset for ploting\n",
      "                        function (deterministic sampling) (default: 100000)\n",
      "  --default_config, -d  Print default configuration file. Can be used to\n",
      "                        generate a template JSON file (default: False)\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Usage examples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Basic usage  (quiet mode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-16T13:15:38.267960Z",
     "start_time": "2020-12-16T13:15:32.382173Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "Loading plotting interface\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f ./data/Albacore-1.2.1_basecall-1D-DNA_sequencing_summary.txt.gz \\\n",
    "    -o ./results/Albacore-1.2.1_basecall-1D-DNA.html \\\n",
    "    --quiet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/pycoQC/results/Albacore-1.2.1_basecall-1D-DNA.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### JSON data output on top of the html report\n",
    "\n",
    "A json report can be generated on top (or instead) of the html report\n",
    "\n",
    "It contains a summarized version of the data collected by pycoQC in a structured and easy to parse format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-16T13:15:46.326442Z",
     "start_time": "2020-12-16T13:15:40.149501Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "Loading plotting interface\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f ./data/Guppy-2.1.3_basecall-1D-RNA_sequencing_summary.txt.gz \\\n",
    "    -o ./results/Guppy-2.1.3_basecall-1D_RNA.html \\\n",
    "    -j ./results/Guppy-2.1.3_basecall-1D_RNA.json \\\n",
    "    --quiet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/pycoQC/results/Guppy-2.1.3_basecall-1D_RNA.html)\n",
    "\n",
    "[JSON OUTPUT](https://a-slide.github.io/pycoQC/pycoQC/results/Guppy-2.1.3_basecall-1D_RNA.json)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Including guppy barcoding information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-16T13:15:46.328458Z",
     "start_time": "2020-12-16T13:15:43.632Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "pycoQC \\\n",
    "    -f ./data/Guppy-2.1.3_basecall-1D-DNA_sequencing_summary.txt.gz \\\n",
    "    -b ./data/Guppy-2.1.3_basecall-1D_DNA_barcoding_summary.txt.gz \\\n",
    "    -o ./results/Guppy-2.1.3_basecall-1D_DNA_barcode.html \\\n",
    "    --quiet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/pycoQC/results/Guppy-2.1.3_basecall-1D_DNA_barcode.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Matching multiple files with a regex and add a title to report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-16T13:15:46.329661Z",
     "start_time": "2020-12-16T13:15:45.948Z"
    },
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "pycoQC \\\n",
    "    -f ./data/Albacore*RNA* \\\n",
    "    -o ./results/Albacore_all_RNA.html \\\n",
    "    --report_title \"All RNA runs\" \\\n",
    "    --quiet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/pycoQC/results/Albacore_all_RNA.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Tweak filtering parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Define reads with a quality higher than 8 and length higher than 200 bases as \"pass\" \n",
    "* Discard reads aligned on the calibration standard\n",
    "* Unset value of any barcode found in less than 10% of the reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-16T13:15:58.130952Z",
     "start_time": "2020-12-16T13:15:51.882778Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "Loading plotting interface\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f ./data/Albacore-2.1.10_basecall-1D-DNA_sequencing_summary.txt.gz\\\n",
    "    -o ./results/Albacore-2.1.10_basecall-1D-DNA.html \\\n",
    "    --min_pass_qual 8 \\\n",
    "    --min_pass_len 200 \\\n",
    "    --filter_calibration \\\n",
    "    --min_barcode_percent 10 \\\n",
    "    --quiet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/pycoQC/results/Albacore-2.1.10_basecall-1D-DNA.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Including Alignments information for a Bam file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-12-16T13:16:10.215380Z",
     "start_time": "2020-12-16T13:15:58.819352Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "Loading plotting interface\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f ./large_data/sample_1_sequencing_summary.txt \\\n",
    "    -a ./large_data/sample_1.bam \\\n",
    "    -o ./results/Guppy-2.3_basecall-1D_alignment-DNA.html \\\n",
    "    -j ./results/Guppy-2.3_basecall-1D_alignment-DNA.json \\\n",
    "    --quiet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/pycoQC/results/Guppy-2.3_basecall-1D_alignment-DNA.html)\n",
    "\n",
    "[JSON OUTPUT](https://a-slide.github.io/pycoQC/pycoQC/results/Guppy-2.3_basecall-1D_alignment-DNA.json)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Advanced configuration with custon json file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Although we recommend to stick to the default parameters, a json formatted configuration file can be provided to tweak the plots. A default configuration file can be generated using:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:12:46.443431Z",
     "start_time": "2020-02-10T14:12:44.120118Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"run_summary\": {\n",
      "    \"plot_title\": \"General run summary\"\n",
      "  },\n",
      "  \"basecall_summary\": {\n",
      "    \"plot_title\": \"Basecall summary\"\n",
      "  },\n",
      "  \"alignment_summary\": {\n",
      "    \"plot_title\": \"Alignment summary\"\n",
      "  },\n",
      "  \"read_len_1D\": {\n",
      "    \"plot_title\": \"Basecalled reads length\",\n",
      "    \"color\": \"lightsteelblue\",\n",
      "    \"nbins\": 200,\n",
      "    \"smooth_sigma\": 2\n",
      "  },\n",
      "  \"align_len_1D\": {\n",
      "    \"plot_title\": \"Aligned reads length\",\n",
      "    \"color\": \"mediumseagreen\",\n",
      "    \"nbins\": 200,\n",
      "    \"smooth_sigma\": 2\n",
      "  },\n",
      "  \"read_qual_1D\": {\n",
      "    \"plot_title\": \"Basecalled reads PHRED quality\",\n",
      "    \"color\": \"salmon\",\n",
      "    \"nbins\": 200,\n",
      "    \"smooth_sigma\": 2\n",
      "  },\n",
      "  \"identity_freq_1D\": {\n",
      "    \"plot_title\": \"Aligned reads identity\",\n",
      "    \"color\": \"sandybrown\",\n",
      "    \"nbins\": 200,\n",
      "    \"smooth_sigma\": 2\n",
      "  },\n",
      "  \"read_len_read_qual_2D\": {\n",
      "    \"plot_title\": \"Basecalled reads length vs reads PHRED quality\",\n",
      "    \"x_nbins\": 200,\n",
      "    \"y_nbins\": 100,\n",
      "    \"smooth_sigma\": 2\n",
      "  },\n",
      "  \"read_len_align_len_2D\": {\n",
      "    \"plot_title\": \"Basecalled reads length vs alignments length\",\n",
      "    \"x_nbins\": 200,\n",
      "    \"y_nbins\": 100,\n",
      "    \"smooth_sigma\": 1\n",
      "  },\n",
      "  \"align_len_identity_freq_2D\": {\n",
      "    \"plot_title\": \"Aligned reads length vs alignments identity\",\n",
      "    \"x_nbins\": 200,\n",
      "    \"y_nbins\": 100,\n",
      "    \"smooth_sigma\": 2\n",
      "  },\n",
      "  \"read_qual_identity_freq_2D\": {\n",
      "    \"plot_title\": \"Reads PHRED quality vs alignments identity\",\n",
      "    \"x_nbins\": 200,\n",
      "    \"y_nbins\": 100,\n",
      "    \"smooth_sigma\": 1\n",
      "  },\n",
      "  \"output_over_time\": {\n",
      "    \"plot_title\": \"Output over experiment time\",\n",
      "    \"cumulative_color\": \"rgb(204,226,255)\",\n",
      "    \"interval_color\": \"rgb(102,168,255)\"\n",
      "  },\n",
      "  \"read_len_over_time\": {\n",
      "    \"plot_title\": \"Read length over experiment time\",\n",
      "    \"median_color\": \"rgb(102,168,255)\",\n",
      "    \"quartile_color\": \"rgb(153,197,255)\",\n",
      "    \"extreme_color\": \"rgba(153,197,255,0.5)\",\n",
      "    \"smooth_sigma\": 1\n",
      "  },\n",
      "  \"read_qual_over_time\": {\n",
      "    \"plot_title\": \"Read quality over experiment time\",\n",
      "    \"median_color\": \"rgb(250,128,114)\",\n",
      "    \"quartile_color\": \"rgb(250,170,160)\",\n",
      "    \"extreme_color\": \"rgba(250,170,160,0.5)\",\n",
      "    \"smooth_sigma\": 1\n",
      "  },\n",
      "  \"align_len_over_time\": {\n",
      "    \"plot_title\": \"Aligned reads length over experiment time\",\n",
      "    \"median_color\": \"rgb(102,168,255)\",\n",
      "    \"quartile_color\": \"rgb(153,197,255)\",\n",
      "    \"extreme_color\": \"rgba(153,197,255,0.5)\",\n",
      "    \"smooth_sigma\": 1\n",
      "  },\n",
      "  \"identity_freq_over_time\": {\n",
      "    \"plot_title\": \"Aligned reads identity over experiment time\",\n",
      "    \"median_color\": \"rgb(250,128,114)\",\n",
      "    \"quartile_color\": \"rgb(250,170,160)\",\n",
      "    \"extreme_color\": \"rgba(250,170,160,0.5)\",\n",
      "    \"smooth_sigma\": 1\n",
      "  },\n",
      " \"barcode_counts\": {\n",
      "    \"plot_title\": \"Number of reads per barcode\",\n",
      "    \"colors\": [\n",
      "      \"#f8bc9c\",\n",
      "      \"#f6e9a1\",\n",
      "      \"#f5f8f2\",\n",
      "      \"#92d9f5\",\n",
      "      \"#4f97ba\"\n",
      "    ]\n",
      "  },\n",
      "  \"channels_activity\": {\n",
      "    \"plot_title\": \"Channel activity over time\",\n",
      "    \"smooth_sigma\": 1\n",
      "  },\n",
      "  \"alignment_reads_status\": {\n",
      "    \"plot_title\": \"Summary of reads alignment\",\n",
      "    \"colors\": [\n",
      "      \"#f8bc9c\",\n",
      "      \"#f6e9a1\",\n",
      "      \"#92d9f5\",\n",
      "      \"#4f97ba\",\n",
      "      \"#f5f8f2\"\n",
      "    ]\n",
      "  },\n",
      "  \"alignment_rate\": {\n",
      "    \"plot_title\": \"Bases alignment rate\",\n",
      "    \"colors\": [\n",
      "      \"#fcaf94\",\n",
      "      \"#828282\",\n",
      "      \"#fc8161\",\n",
      "      \"#828282\",\n",
      "      \"#f44f39\",\n",
      "      \"#d52221\",\n",
      "      \"#828282\",\n",
      "      \"#828282\",\n",
      "      \"#828282\",\n",
      "      \"#828282\"\n",
      "    ]\n",
      "  },\n",
      "  \"alignment_coverage\": {\n",
      "    \"plot_title\": \"Coverage overview\",\n",
      "    \"nbins\": 500,\n",
      "    \"color\": \"rgba(102,168,255,0.75)\",\n",
      "    \"smooth_sigma\": 1\n",
      "  }\n",
      "}\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC --default_config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-06-24T15:52:37.096001Z",
     "start_time": "2019-06-24T15:52:36.888372Z"
    }
   },
   "source": [
    "To save and edit it redirect the std output to a file and make your changes using your favorite text editor.\n",
    "\n",
    "To remove a plot from the report, just remove it (or comment it) from the configuration file\n",
    "\n",
    "The configuration file accept all the arguments of the target plotting functions. For more information refer to the API documentation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-11-07T16:01:43.427091Z",
     "start_time": "2019-11-07T16:01:41.355758Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC --default_config > data/pycoQC_config.json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run pycoQC with `--config` option"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-02-10T14:13:01.620742Z",
     "start_time": "2020-02-10T14:12:54.500540Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking arguments values\n",
      "Check input data files\n",
      "Parse data files\n",
      "Merge data\n",
      "Cleaning data\n",
      "Loading plotting interface\n",
      "(pycoQC) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "pycoQC \\\n",
    "    -f ./data/Albacore-1.7.0_basecall-1D-DNA_sequencing_summary.txt.gz\\\n",
    "    -o ./results/Albacore-1.7.0_basecall-1D-DNA.html \\\n",
    "    --config ./data/pycoQC_config.json \\\n",
    "    --quiet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[HTML OUTPUT](https://a-slide.github.io/pycoQC/pycoQC/results/Albacore-1.7.0_basecall-1D-DNA.html)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  },
  "widgets": {
   "state": {},
   "version": "1.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
